nbtools/run_notebook.py

```diff
@@ -0,0 +1,287 @@
+""" Utility functions for running Jupyter Notebooks."""
+#pylint: disable=import-outside-toplevel
+import os
+import time
+from glob import glob
+from textwrap import dedent
+
+# Code fragments that inserted in the notebook
+CELL_INSERT_COMMENT = "# Cell inserted during automated execution"
+
+# Connect with shelve database for inputs/outputs providing
+DB_CONNECT_CODE_CELL = """
+    import os, shelve
+    from dill import Pickler, Unpickler
+
+    shelve.Pickler = Pickler
+    shelve.Unpickler = Unpickler
+
+    out_path_db = {}
+"""
+DB_CONNECT_CODE_CELL = dedent(DB_CONNECT_CODE_CELL)
+
+# Insert params into the notebook
+INPUTS_CODE_CELL = """
+    # Inputs loading
+    with shelve.open(out_path_db) as notebook_db:
+        inputs = {**notebook_db}
+
+        locals().update(inputs)
+"""
+INPUTS_CODE_CELL = dedent(INPUTS_CODE_CELL)
+
+# Save notebook outputs
+OUTPUTS_CODE_CELL = """
+    # Output dict preparation
+    output = {{}}
+    outputs = {}
+
+    for value_name in outputs:
+        if value_name in locals():
+            output[value_name] = locals()[value_name]
+
+    with shelve.open(out_path_db) as notebook_db:
+        notebook_db['outputs'] = output
+"""
+OUTPUTS_CODE_CELL = dedent(OUTPUTS_CODE_CELL)
+
+
+def run_notebook(path, inputs=None, outputs=None, inputs_pos=1, out_path_db=None, execute_kwargs=None,
+                 out_path_ipynb=None, out_path_html=None, add_timestamp=True, hide_code_cells=False, display_links=True,
+                 raise_exception=False, return_notebook=False):
+    """ Run a notebook and save the execution result.
```

Comment:
Make the documentation a more complete overview: a simple explanation of what is going on, details and pitfals of implementation